Pr1/3 : add GPG key management operations and facts

maisim opened the pull request:
Add new gpg.key and gpg.dearmor operations to manage GPG keys and keyrings. These operations provide a modern alternative to apt-key for managing APT repository keys.

Features:
- Install keys from URLs, local files, or keyservers
- Remove keys by ID or entire keyring files
- Convert ASCII armored keys to binary format
- Manage keys in specific keyrings or across all APT keyrings

This is part 1/3 of modernizing APT key management.

<!--
    🎉 Thank you for taking the time to contribute to pyinfra! 🎉

    Please provide a short description of the proposed change and here's a handy checklist of things
    to make PRs quicker to review and merge.

    Note that we will not merge new connectors, but instead welcome PRs that link to thid party
    connector packages.
-->

- [x] Pull request is based on the default branch (`3.x` at this time)
- [x] Pull request includes tests for any new/updated operations/facts
- [x] Pull request includes documentation for any new/updated operations/facts
- [x] Tests pass (see `scripts/dev-test.sh`)
- [x] Type checking & code style passes (see `scripts/dev-lint.sh`)

maisim commented:
Hi @DonDebonair !
Can you take a look at this, please?
It seems to work, but I'm not sure I'm always using gpg correctly. Also, I'm always hesitant to use facts or shell commands within operations. Let me know what you think.
Thanks!

DonDebonair commented:
Thanks for the PR @maisim . I would not be afraid to use facts inside operations. I actually consider it a best practice to use facts to check things in operations. So instead of doing checks inside the shell commands you yield, there are probably places where you can rely on facts instead. That makes the yielded commands simpler.

If you can change those already, I'll do a more full review this weekend.

Caveat btw: I'm by no means a GPG expert. A short while ago, I wanted to use pyinfra to install Docker on a Debian host. I just created the keyring directory and downloaded the key directly into that, never touching GPG 😅

DonDebonair commented on src/pyinfra/operations/gpg.py:
```diff
@@ -0,0 +1,300 @@
+                for kid in keyid:
+                    # Remove key from all matching keyrings
+                    yield (
+                        f'for keyring in {pattern}; do [ -e "$keyring" ] && '
```
Doesn't it make more sense to retrieve keyrings through a fact and then only try to remove the key from keyrings that actually exist?

Or maybe even better, create a fact that lists all known keyrings and then use the `GpgKeys` fact for each of those keyrings to get all the keys. Then you can specifically remove the keyid if it exists

maisim replied:
Good idea. Since there’s no global or common location for keyrings, this will allow specifying the directories to search in — for example, /etc/apt/trusted.gpg.d/ and /usr/share/keyrings/ in the APT context.

DonDebonair commented on src/pyinfra/operations/gpg.py:
```diff
@@ -0,0 +1,300 @@
+
+                # Clean up empty keyrings
+                yield (
+                    f'for keyring in {pattern}; do [ -e "$keyring" ] && '
```
As referenced in the other comment, if you can list all keys by keyring, then you know if keyrings are going to be empty after removing `keyid` from those rings and you can remove them based on that

DonDebonair commented on src/pyinfra/operations/gpg.py:
```diff
@@ -0,0 +1,300 @@
+            yield (f'if [ -f "{dest}" ] && ({condition}); then ' f'rm -f "{dest}"; fi')
+            return
+
+        elif dest and not keyid:
```
This could be `else` instead of `elif`. To me, that seems cleaner because it clearly signals there are only 3 possible combinations of conditions: `keyid` is provided and `dest` isn't, `keyid` and `dest` are both provided and lastly, `dest` is provided and `keyid` isn't

maisim replied:
Yes, but it is more an "documentation" elif. To be more explicit.
What about add an else wich raise an exception if we go into an unhandle case?

DonDebonair replied:
That works for me!

DonDebonair commented on src/pyinfra/operations/gpg.py:
```diff
@@ -0,0 +1,300 @@
+        elif dest and keyid:
+            # For APT keyring files, use a simpler approach:
+            # Check if keys exist in file, and if so, remove the entire file
+            # This is appropriate for most APT use cases
```
I'm a bit concerned about "most APT use cases" here. Is there a situation we might be removed a keyfile/keyring that has multiple keys in it?

DonDebonair commented on src/pyinfra/operations/gpg.py:
```diff
@@ -0,0 +1,300 @@
+
+
+@operation()
+def key(
```
This is a very complex function that covers many different scenarios. I think that from a public API standpoint, that is fine, because we're dealing with 1 resource: keys. But you could break up the function into separate functions that are all called from the `key()` function, based on the scenario.

See the `docker` operations for an example of how this can be done

DonDebonair commented:
Left some more comments @maisim incl. some ideas on how to leverate facts.

maisim committed:
operations/facts: add GPG key management operations and facts

- Add GpgFactBase, GpgKey, GpgKeys, GpgSecretKeys, and GpgKeyrings facts
- Add gpg.key and gpg.dearmor operations for managing GPG keys
- Support for keyserver fetching, local files, URLs, and key removal
- Comprehensive test coverage for all GPG operations and facts

maisim committed:
operations/gpg: refactor key() function for better maintainability

- Break down complex key() function into smaller, focused helper functions
- Add _install_key_from_src() for file/URL installation
- Add _install_key_from_keyserver() for keyserver installation
- Add _remove_key_from_keyrings() for multi-keyring removal
- Add _remove_key_from_keyring() for single keyring removal
- Add _remove_keyring_file() for complete keyring removal
- Add validation helper functions for parameters
- Improve code readability and maintainability
- Follow Docker operations pattern for function organization
- Maintain 100% backward compatibility with existing API

maisim committed:
typos: exclude GPG keyring test files from spell checking

GPG key IDs contain hexadecimal values that trigger false positives
in typos (e.g., 'BA' in '1E9377A2BA9EF27F' is flagged as misspelled).
These are valid hex values, not typos.

maisim committed:
feat(gpg): add idempotent key installation check and new test case

Fizzadar commented on src/pyinfra/facts/gpg.py:
```diff
@@ -148,3 +148,71 @@ def command(self, keyring=None):
+            f"for keyring in $(find {search_locations} -type f \\( -name '*.gpg' "
+            f"-o -name '*.asc' -o -name '*.kbx' \\) 2>/dev/null); do "
+            f'echo "KEYRING:$keyring"; '
+            f'if [[ "$keyring" == *.asc ]]; then '
```
This check feels risky as it's bash, not sh/posix, compatible. Two options I can think of:

- split the script to run find twice, you could also then just `find -exec` the gpg call after which would be cleaner, just do one for `*.gpg, *.kbx` and a separate for `.asc`
- use `case "$keyring" in *.asc)` or similar

maisim replied:
You're right
done in a9f02f6b7cfa196b94bfefdf04bbcace48f82773
I used the find exec solution.

Fizzadar commented on src/pyinfra/operations/gpg.py:
```diff
@@ -0,0 +1,414 @@
+    joined = " ".join(keyid)
+
+    # Create temporary GPG home directory
+    temp_dir = f"/tmp/pyinfra-gpg-{host.get_temp_filename('')[-8:]}"
```
We should use `host.get_temp_filename` here.

maisim replied:
:+1: done in 66580a7b5c6f0cc209c5d14a75ef7498b2089880

Fizzadar commented on src/pyinfra/operations/gpg.py:
```diff
@@ -0,0 +1,414 @@
+                    keys_to_remove.append(existing_key_id)
+
+        if keys_to_remove:
+            # Remove the entire keyring file if any target keys are found
```
This seems very wrong to me! Why would we want to nuke the entire file? Surly we'd only do that if _all_ keys match for deletion?

`gpg --batch --delete-keys` would probably be more appropriate in all cases here.

maisim replied:
At the beginning, 23a22345, I used --delete-keys. In 04349890, I did a change to remove the whole file...
Cannot remember why I did that. Maybe beacause we cannot remove key from armored keyrings... 
Now, on only a key is removed, if there no more key in the keyring, the file is deleted.
Fom armored keyrings, if the key have to be removed but the keyring contains another one, it raise an OperationError

Let me know

Fizzadar replied:
I think this is perfect, would much rather error in the armored case than remove the whole file 👍

Fizzadar commented on src/pyinfra/operations/gpg.py:
```diff
@@ -0,0 +1,414 @@
+                break
+
+        if keys_found:
+            # Remove the entire keyring file - safest approach for keyring management
```
Same as above!

Fizzadar reviewed (changes requested):
Hi @maisim! Apologies for the big delay here, picking this up while Daan is afk, added a few things to fix.

maisim committed:
refactor(gpg): update keyring command to use find with exec for improved compatibility

maisim committed:
refactor(gpg): use built-in temporary directory naming for keyserver operations

maisim committed:
feat(gpg): enhance key removal operations with individual key deletion and error handling for armored keyrings

maisim committed:
fix: _matches_keyid handles full 40-char fingerprint vs short key ID

GpgKeyrings stores the 16-char short key ID (bits[4] from the gpg pub
record). When the caller supplies a full 40-char fingerprint, the short ID
is always its last 16 hex digits, so add a clean_key.endswith(existing_upper)
check alongside the existing conditions.

maisim committed:
test: add fixture for gpg.key removal by full 40-char fingerprint

Covers the clean_key.endswith(existing_upper) path in _matches_keyid:
the keyring stores the 16-char short key ID (last 16 hex digits of the
fingerprint) while the caller supplies the full 40-char fingerprint.

Fizzadar reviewed (approved):
Very nice, @maisim!
